File: backend/app/api/routes/articles.py
import uuid

from fastapi import APIRouter, HTTPException
from sqlalchemy.orm import joinedload

from ... import crud
from ...models import Article, ArticleCreate, ArticleRead, ArticleUpdate
from ..deps import CurrentUser, SessionDep

router = APIRouter()


@router.post("/", response_model=ArticleRead)
def create_article(
    *, session: SessionDep, article_create: ArticleCreate, current_user: CurrentUser
):
    return crud.create_article(
        session=session, article_create=article_create, user=current_user
    )


@router.get("/", response_model=list[ArticleRead] | ArticleRead)
def get_articles(
    *,
    session: SessionDep,
    current_user: CurrentUser,
    article_id: uuid.UUID | None = None,
    author_id: uuid.UUID | None = None,
):
    if article_id and author_id:
        raise HTTPException(
            status_code=400, detail="Cannot filter by both article and author"
        )
    elif article_id:
        article = crud.get_article_by_id(session=session, article_id=article_id)
        if article:
            return article
        raise HTTPException(status_code=404, detail="Article not found")
    elif author_id:
        return crud.get_articles_by_author_id(session=session, author_id=author_id)
    return crud.get_articles(session=session)


# 47034041-62ea-4ef1-ab83-d83c24a42906
@router.put("/", response_model=ArticleRead)
def edit_article(
    *,
    session: SessionDep,
    current_user: CurrentUser,
    article_id: uuid.UUID,
    article_update: ArticleUpdate,
):
    article = crud.get_article_by_id(session=session, article_id=article_id)
    if article:
        if current_user.is_superuser or current_user.id == article.author_id:
            return crud.update_article(
                session=session, db_article=article, article_update=article_update
            )
        raise HTTPException(status_code=403, detail="You are not the author")
    raise HTTPException(status_code=404, detail="Article not found")


@router.delete("/", response_model=ArticleRead)
def delete_article(
    *, session: SessionDep, current_user: CurrentUser, article_id: uuid.UUID
):
    article = crud.get_article_by_id(session=session, article_id=article_id)
    if article:
        if current_user.is_superuser or current_user.id == article.author_id:
            return crud.delete_article(session=session, db_article=article)
        raise HTTPException(status_code=403, detail="You are not the author")
    raise HTTPException(status_code=404, detail="Article not found")

File: frontend/src/api/articles.ts
import api from "../utils";
import { ArticleCreate, ArticleRead, ArticleUpdate } from "../types/Article";

export const articlesApi = {
  createArticle: async (articleCreate: ArticleCreate): Promise<ArticleRead> => {
    const response = await api.post<ArticleRead>("/articles", articleCreate, {
      headers: {
        "Content-Type": "application/json",
      },
    });
    return response.data;
  },
  getArticles: async (): Promise<ArticleRead[]> => {
    const response = await api.get<ArticleRead[]>("/articles");
    return response.data;
  },
  getArticlesByAuthorId: async (authorId: string): Promise<ArticleRead[]> => {
    const response = await api.get<ArticleRead[]>(
      "/articles/?author_id=" + authorId
    );
    return response.data;
  },
  getArticleById: async (articleId: string): Promise<ArticleRead> => {
    const response = await api.get<ArticleRead>(
      "/articles/?article_id=" + articleId
    );
    return response.data;
  },
  updateArticle: async (
    articleId: string,
    articleUpdate: ArticleUpdate
  ): Promise<ArticleRead> => {
    const response = await api.put<ArticleRead>(
      `/articles/?article_id=${articleId}`,
      articleUpdate,
      {
        headers: {
          "Content-Type": "application/json",
        },
      }
    );
    return response.data;
  },
  deleteArticle: async (articleId: string): Promise<void> => {
    const response = await api.delete(`/articles/?article_id=${articleId}`);
    return response.data;
  },
};
